File: micro-manager/web-manager/src/components/metrics/ContainerSimulatedMetrics.js
/*
 * MIT License
 *
 * Copyright (c) 2020 micro-manager
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React from 'react';
import M from 'materialize-css';
import Utils from '../../utils';
import ContainerSimulatedMetricsCard from './ContainerSimulatedMetricsCard';
import MainLayout from '../shared/MainLayout';
import {Link} from 'react-router-dom';

export default class ContainerSimulatedMetrics extends React.Component {
  constructor (props) {
    super(props);
    this.state = { data: [], loading: false, showAdd: true };
  }

  componentDidMount = () => {
    this.loadSimulatedMetrics();
    const instances = M.Tooltip.init(document.querySelectorAll('.tooltipped'));
    this.setState({ tooltipInstances: instances });
  };

  componentWillUnmount = () => {
    this.state.tooltipInstances[0].destroy();
  };

  loadSimulatedMetrics = () => {
    this.setState({ loading: true });
    Utils.ajaxGet(
      'localhost/metrics/simulated/containers',
      (data) => this.setState({ data: data, loading: false })
    );
  };

  renderSimulatedMetrics = () => {
    let simulatedMetricsNodes;
    if (this.state.data) {
      simulatedMetricsNodes = this.state.data.map((simulatedMetric, index) => (
        <ContainerSimulatedMetricsCard key={simulatedMetric.id} simulatedMetric={simulatedMetric}
                                       reloadSimulatedMetrics={this.loadSimulatedMetrics}/>
      ));
    }
    return simulatedMetricsNodes;
  };

  render = () => (
    <MainLayout title='Container simulated metrics'>
      {this.renderSimulatedMetrics()}
      <div className="fixed-action-btn tooltipped" data-position="left" data-tooltip="Add container simulated metric">
        <Link className="waves-effect waves-light btn-floating btn-large grey darken-4"
              to='/ui/simulatedmetrics/containers/detail'>
          <i className="large material-icons">add</i>
        </Link>
      </div>
    </MainLayout>
  );
}

File: usmanager/manager/manager-database/src/main/java/pt/unl/fct/miei/usmanagement/manager/metrics/simulated/ContainerSimulatedMetrics.java
/*
 * MIT License
 *
 * Copyright (c) 2020 manager
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package pt.unl.fct.miei.usmanagement.manager.metrics.simulated;

import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import pt.unl.fct.miei.usmanagement.manager.containers.Container;

import java.util.List;
import java.util.Optional;

public interface ContainerSimulatedMetrics extends JpaRepository<ContainerSimulatedMetric, Long> {

	Optional<ContainerSimulatedMetric> findByNameIgnoreCase(@Param("name") String name);

	@Query("select m "
		+ "from ContainerSimulatedMetric m left join m.containers c "
		+ "where c.id = :containerId")
	List<ContainerSimulatedMetric> findByContainer(@Param("containerId") String containerId);

	@Query("select m "
		+ "from ContainerSimulatedMetric m left join m.containers c "
		+ "where c.id = :containerId and lower(m.field.name) = lower(:field)")
	Optional<ContainerSimulatedMetric> findByContainerAndField(@Param("containerId") String containerId,
															   @Param("field") String field);

	@Query("select case when count(m) > 0 then true else false end "
		+ "from ContainerSimulatedMetric m "
		+ "where lower(m.name) = lower(:simulatedMetricName)")
	boolean hasContainerSimulatedMetric(@Param("simulatedMetricName") String simulatedMetricName);

	@Query("select c "
		+ "from ContainerSimulatedMetric m join m.containers c "
		+ "where lower(m.name) = lower(:simulatedMetricName)")
	List<Container> getContainers(@Param("simulatedMetricName") String simulatedMetricName);

	@Query("select c "
		+ "from ContainerSimulatedMetric m join m.containers c "
		+ "where lower(m.name) = lower(:simulatedMetricName) and c.id = :containerId")
	Optional<Container> getContainer(@Param("simulatedMetricName") String simulatedMetricName,
									 @Param("containerId") String containerId);


}
